File: FourBar_PositionAnalysisNewtonMethod.py
"""
Four-Bar Analysis using Newton Iteration (Numerical Solver)
Combined from FourBar_PositionAnalysisNewtonMethod.m and vecloopeq.m
Solves the kinematics equations of a planar 4-bar linkage using fsolve (Newton's method)
"""
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import fsolve

# Link lengths and parameters
r1 = 1
r2 = 2
r3 = 3.5
r4 = 4
th1 = 0
beta = np.arctan(1/2)  # coupler angle
r6 = 2 / np.cos(beta)   # length AE

th2list = np.deg2rad(np.arange(0, 361, 1))  # driver angles
n = len(th2list)

th3list = np.zeros(n)
th4list = np.zeros(n)
pAlist = np.zeros((n, 2))
pBlist = np.zeros((n, 2))
pElist = np.zeros((n, 2))

OB = np.array([1, 0])

# Vector loop equation for 4-bar linkage
def vecloopeq(x, th1, th2, r1, r2, r3, r4):
    th3, th4 = x
    eq1 = r2 * np.cos(th2) + r3 * np.cos(th3) - r4 * np.cos(th4) - r1 * np.cos(th1)
    eq2 = r2 * np.sin(th2) + r3 * np.sin(th3) - r4 * np.sin(th4) - r1 * np.sin(th1)
    return [eq1, eq2]

# Initial guess for [th3, th4]
th2 = th2list[0]
x0 = [-np.pi/3, -np.pi/3]

for i, th2 in enumerate(th2list):
    f = lambda x: vecloopeq(x, th1, th2, r1, r2, r3, r4)
    solth3th4 = fsolve(f, x0)
    th3, th4 = solth3th4
    th3list[i] = th3
    th4list[i] = th4
    x0 = solth3th4  # use previous solution as next initial guess

    pA = r2 * np.array([np.cos(th2), np.sin(th2)])
    pB = OB + r4 * np.array([np.cos(th4), np.sin(th4)])
    pE = pA + r6 * np.array([np.cos(th3 + beta), np.sin(th3 + beta)])

    pAlist[i] = pA
    pBlist[i] = pB 
    pElist[i] = pE

plt.figure()
plt.plot(th2list, th3list, label='Coupler Angle th3')
plt.plot(th2list, th4list, label='Output Angle th4')
plt.xlabel('Crank Angle (rad)')
plt.figure()
plt.plot(pElist[:,0], pElist[:,1], 'r-', label='Coupler Point E Trajectory')
plt.xlabel('X Position')
plt.ylabel('Y Position')
plt.title('Trajectory of Coupler Point E')
plt.axis('equal')
plt.grid(True)
plt.legend()
plt.ylabel('Angle (rad)')
plt.title('Coupler and Output Angles vs. Input Crank Angle')
plt.grid(True)
plt.legend()
plt.show()

File: FourBar_PositionAnalysisNewtonMethod.m
% Four-Bar Analysis
% Use Newton Iteration approch
clear all;
clc;
close all;

r1=1;
r2=2;
r3=3.5;
r4=4;
th1=0;
beta=atan(1/2); %coupler angle
r6=2/cos(beta); %length AE

th2list=(0:1:360)*pi/180; %driver
n=length(th2list);

th3list=zeros(n,1);
th4list=zeros(n,1);
pAlist=zeros(n,2);
pBlist=zeros(n,2);
pElist=zeros(n,2);

th2=0;
x0=[-pi/3 -pi/3];

%th2=2*pi/3
%x0=[pi/6 pi/3]

f = @(x) vecloopeq(x,th1,th2,r1,r2,r3,r4);
solth3th4=fsolve(f,x0);
th3=solth3th4(1);
th4=solth3th4(2);
pA=r2*[cos(th2) sin(th2)]
pE=pA+2*[cos(th3) sin(th3)]+1*[cos(th3+pi/2) sin(th3+pi/2)]
OA=[0 0]
OB=[1 0]

for i = 1:length(th2list)
    th2=th2list(i);
    f = @(x) vecloopeq(x,th1,th2,r1,r2,r3,r4);
    solth3th4=fsolve(f,x0);
    th3=solth3th4(1);
    th4=solth3th4(2);
    th3list(i)=th3;
    th4list(i)=th4;
    x0=solth3th4;
    
    pAlist(i,:)=r2*[cos(th2) sin(th2)];
    pBlist(i,:)=OB+r4*[cos(th4) sin(th4)];
    pElist(i,:)=pAlist(i,:)+r6*[cos(th3+beta) sin(th3+beta)];
    
end
 
plot(th2list,th3list,th2list,th4list)
xlabel('Crank Angle')
ylabel('Coupler and Output Angle')
title('Coupler and Output Angles vs. Input Crank Angle')
grid on
set(gca,'XTick')


